File: frogment/src/main/java/ch/grze/frogment/activity/AbstractFrogmentActivity.java
package ch.grze.frogment.activity;

import android.support.v7.app.AppCompatActivity;

//still in Java due to lack of interop: https://youtrack.jetbrains.com/issue/KT-19415
public abstract class AbstractFrogmentActivity extends AppCompatActivity implements FrogmentActivityInterface {
    private FrogmentActivityComponent frogmentActivityComponent;

    @Override
    public FrogmentActivityComponent getFrogmentActivityComponent() {
        return frogmentActivityComponent;
    }

    @Override
    public void setFrogmentActivityComponent(FrogmentActivityComponent component) {
        frogmentActivityComponent = component;
    }
}

File: frogment/src/main/java/ch/grze/frogment/activity/AbstractFrogmentActivity.kt
package ch.grze.frogment.activity

import android.support.v7.app.AppCompatActivity
import ch.grze.frogment.core.backstack.BackStackChangeListener
import ch.grze.frogment.core.navigation.NavigatorAware

abstract class AbstractFrogmentActivity private constructor(override val frogmentActivityComponent: FrogmentActivityComponent) :
        AppCompatActivity(),
        FrogmentActivityInterface,
        NavigatorAware by frogmentActivityComponent,
        BackStackChangeListener by frogmentActivityComponent {

    constructor() : this(FrogmentActivityComponent())

    init {
        frogmentActivityComponent.frogmentActivity = this
    }
}
